operator/controllers/remote.go

```diff
@@ -0,0 +1,187 @@
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"github.com/kyma-project/manifest-operator/api/api/v1alpha1"
+	"github.com/kyma-project/manifest-operator/operator/pkg/custom"
+	"github.com/kyma-project/manifest-operator/operator/pkg/labels"
+	"github.com/kyma-project/manifest-operator/operator/pkg/manifest"
+	v1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"path"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+type RemoteInterface struct {
+	NativeClient client.Client
+	CustomClient client.Client
+	NativeObject *v1alpha1.Manifest
+}
+
+func NewRemoteInterface(ctx context.Context, nativeClient client.Client, nativeObject *v1alpha1.Manifest) (*RemoteInterface, *v1alpha1.Manifest, error) {
+	namespacedName := client.ObjectKeyFromObject(nativeObject)
+	kymaOwner, ok := nativeObject.Labels[labels.ComponentOwner]
+	if !ok {
+		return nil, nil, fmt.Errorf(
+			fmt.Sprintf("label %s not found on manifest resource %s", labels.ComponentOwner, namespacedName))
+	}
+
+	remoteCluster := &custom.ClusterClient{DefaultClient: nativeClient}
+	restConfig, err := remoteCluster.GetRestConfig(ctx, kymaOwner, nativeObject.Namespace)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	customClient, err := remoteCluster.GetNewClient(restConfig, client.Options{
+		Scheme: nativeClient.Scheme(),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	remoteInterface := &RemoteInterface{
+		NativeObject: nativeObject,
+		NativeClient: nativeClient,
+		CustomClient: customClient,
+	}
+
+	if meta.IsNoMatchError(err) {
+		if err := remoteInterface.CreateCRD(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check for remote resource
+	remoteManifest, err := remoteInterface.GetLatestRemote(ctx)
+
+	if client.IgnoreNotFound(err) != nil {
+		// unknown error
+		return nil, nil, err
+	} else if err != nil {
+		// remote resource doesn't exist
+		remoteManifest, err = remoteInterface.createRemote(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check finalizer
+	if !controllerutil.ContainsFinalizer(remoteManifest, labels.ManifestFinalizer) {
```

Comment:
I think the finalizer would already have been created in the create remote. One of those checks or the assignment of the finalizer in the creation is redundant here I think

Reply:
It's just a check that we are consistent like the native object. Making sure the user hasn't accidentally removed it.

Reply:
That I understand but in line 66 you createRemote and there you also ensure the finalizer. Im saying that if you have this logic, the createRemote doesnt need to contain the finalizer, so theres always only one place where its added